### PR TITLE
BUG: protect against accessing base attribute of a NULL subarray

### LINKED ISSUE #19325: dtype-related segfault when taking a view in NumPy 1.21
I am seeing a segfault in the zarr-python test suite when using NumPy 1.21. It appears to be related to taking a `view` of an array with void type:

### Reproducing code example:
Here is a small reproducer independent of Zarr:

```Python
import numpy as np

dtype = np.dtype([('foo', '<i8'), ('bar', [('foo', '<i4'), ('bar', '<f4', (3, 2)), ('baz', 'u1')]), ('baz', 'S3')])

x = np.array(b'', dtype=dtype.str)

x.view(dtype)  # segfault
```

In 1.20.3 the output of the code above is:

```
array((0, (0, [[0., 0.], [0., 0.], [0., 0.]], 0), b''),
      dtype=[('foo', '<i8'), ('bar', [('foo', '<i4'), ('bar', '<f4', (3, 2)), ('baz', 'u1')]), ('baz', 'S3')])
```
### Error message:

```
thread 1 "python" received signal SIGSEGV, Segmentation fault.
void_to_void_resolve_descriptors (self=0x7fffc50451b0, dtypes=0x7fffffffc120, given_descrs=0x7fffffffc090, loop_descrs=0x7fffffffc0a0) at numpy/core/src/multiarray/convert_datatype.c:3296
3296	        NPY_CASTING field_casting = PyArray_GetCastSafety(
```

### NumPy/Python version information:

1.21.0 3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37) 
[GCC 9.3.0]

### PR DESCRIPTION
**Pull request by grlee77**
closes #19325

This PR is to protect against potential segfaults in `void_to_void_resolve_descriptors`. This is the first time I have looked at this code, so there may be a better solution, but this seems to work for the example given in #19325

### CODE REVIEW

**Comment by grlee77** [comment]
If this seems like a reasonable fix, we can add a new test case for it

**Comment by grlee77** [comment]
#19322 makes other changes in the same file, but does not resolve the issue addressed here

File: numpy/core/src/multiarray/convert_datatype.c [hunk 1]
```diff
@@ -3293,8 +3296,10 @@ void_to_void_resolve_descriptors(
                 casting = NPY_NO_CASTING | _NPY_CAST_IS_VIEW;
             }
         }
-        NPY_CASTING field_casting = PyArray_GetCastSafety(
-                given_descrs[0]->subarray->base, given_descrs[1]->subarray->base, NULL);
+
+        PyArray_Descr *from_base = (from_sub == NULL) ? NULL : from_sub->base;
+        PyArray_Descr *to_base = (to_sub == NULL) ? NULL : to_sub->base;
+        NPY_CASTING field_casting = PyArray_GetCastSafety(from_base, to_base, NULL);
```
Comment (seberg) [inline numpy/core/src/multiarray/convert_datatype.c:3302]: This is the right idea, but should be `given_descrs[0]` or `given_descrs[1]` rather than `NULL`.  I guess the unstructured `void` case is missing (i.e. there must be a `V4` or similar involved here somehow), I don't know yet how or why `view` even triggers it...

For tests, it may be good to use also check with `np.can_cast` (`casting="unsafe"`).

Comment (seberg) [reply numpy/core/src/multiarray/convert_datatype.c:3302]: Ah, I had not realized that in your example `x.dtype` is `"V40"`, than it all adds up.

Comment (grlee77) [reply numpy/core/src/multiarray/convert_datatype.c:3302]: Sure, that works without requiring any change to `PyArray_GetCastSafety` itself. Let me look into a test case and I will update

**Comment by grlee77** [comment]
The following is a minimal `can_cast` call that segfaults on 1.21.0 and works for this PR (It returns True if I set casting="unsafe", otherwise False)
```Python
np.can_cast('V40', np.dtype([('foo', '<f4', (3, 2))]), casting="unsafe")
```

Cases not involving the shape argument in the dtype seem fine on 1.21.0, e.g.
```Python
np.can_cast(
    'V40', np.dtype([('foo', '<i8'), ('bar', [('foo', '<i4'), ('baz', 'u1')])]),  casting="unsafe"
)
```
does not segfault

**Comment by seberg** [comment]
Thanks, this looks good to me.  If you are up to it, adding a test for the reverse direction would be good (I know its even stranger).

I am still considering if there may be some incorrect corner cases for `subarray -> void` casts.  The strangest thing that works is probably:
```
arr = np.ones(3, dtype="i4,(1,2)i4")
arr.astype("i4,V4")
```
(The crazy part is that we the first element of the subarray, ignoring the rest!  I don't want to backport such a change, but maybe we should consider that a bug...  If you change it to `V8` there, you do not get the second `i4` at all)

**Review by seberg — APPROVED** [review]
Thanks, this works for me.  There are probably more subtleties we should test, but there is no pressing reason to do it here.

**Comment by charris** [comment]
Thanks @grlee77 .
